c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -0,0 +1,506 @@
+use crate::rewrite::Rewrite;
+use rustc_hir::Mutability;
+use rustc_span::source_map::{FileName, SourceMap};
+use rustc_span::{BytePos, SourceFile, Span, SyntaxContext};
+use std::cmp::Reverse;
+use std::collections::HashMap;
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum RewriteError<S = Span> {
+    /// The provided rewrite overlaps, but is not contained in, another rewrite.  `.0` is the span
+    /// of the other rewrite.
+    PartialOverlap(S),
+    /// The provided rewrite conflicts with a different rewrite at the same span.
+    Conflict,
+    /// The provided rewrite affects code that would be discarded by a rewrite of a containing
+    /// expression.  `.0` is the span of the containing expression, and `.1` is its rewrite.
+    _Discarded(S, Box<Rewrite<S>>),
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+struct RewriteTree<S = Span> {
+    span: S,
+    rw: Rewrite<S>,
+    /// Child nodes, representing rewrites that are contained entirely within the current rewrite.
+    /// Each child's `span` is contained within `self.span`, no child's `span` overlaps any other
+    /// child's `span`, and children are sorted by position (`span.lo()`).
+    children: Vec<RewriteTree<S>>,
+}
+
+trait SpanLike: Copy + Eq {
+    fn lo(self) -> BytePos;
+    fn hi(self) -> BytePos;
+    fn contains(self, other: Self) -> bool;
+    fn overlaps(self, other: Self) -> bool;
+}
+
+impl SpanLike for Span {
+    fn lo(self) -> BytePos {
+        self.lo()
+    }
+    fn hi(self) -> BytePos {
+        self.hi()
+    }
+    fn contains(self, other: Self) -> bool {
+        self.contains(other)
+    }
+    fn overlaps(self, other: Self) -> bool {
+        self.overlaps(other)
+    }
+}
+
+impl SpanLike for (u32, u32) {
+    fn lo(self) -> BytePos {
+        BytePos(self.0)
+    }
+    fn hi(self) -> BytePos {
+        BytePos(self.1)
+    }
+    fn contains(self, other: Self) -> bool {
+        self.0 <= other.0 && other.1 <= self.1
+    }
+    fn overlaps(self, other: Self) -> bool {
+        self.0 < other.1 && other.0 < self.1
+    }
+}
+
+impl<S: SpanLike> RewriteTree<S> {
+    pub fn build(
+        mut rws: Vec<(S, Rewrite<S>)>,
+    ) -> (Vec<RewriteTree<S>>, Vec<(S, Rewrite<S>, RewriteError<S>)>) {
+        // Sort by start position and then by decreasing length, so that each parent span comes
+        // before all its children.
+        rws.sort_by_key(|&(ref s, _)| (s.lo(), Reverse(s.hi() - s.lo())));
+
+        // The stack contains a `RewriteTree` node for the most recently processed item from `rws`
+        // along with all of its ancestors (nodes whose spans strictly contain its span).
```

Comment:
i thought the parent's span completely contained the child's span? to me, this comment makes it seem like it's the other way around. which is correct?

Reply:
You have it right.  Ancestors are "nodes whose spans strictly contain [the item's] span".